xarray/coding/variables.py

```diff
@@ -0,0 +1,194 @@
+            'how a variable is serialized. To proceed, remove this key from '
+            "the variable's attributes manually.".format(key, var_str))
+    dest[key] = value
+
+
+def pop_to(source, dest, key, name=None):
+    """
+    A convenience function which pops a key k from source to dest.
+    None values are not passed on.  If k already exists in dest an
+    error is raised.
+    """
+    value = source.pop(key, None)
+    if value is not None:
+        safe_setitem(dest, key, value, name=name)
+    return value
+
+
+def _apply_mask(data,  # type: np.ndarray
+                encoded_fill_values,  # type: list
+                decoded_fill_value  # type: Any
+                ):  # type: npndarray
+    """Mask all matching values in a NumPy arrays."""
+    condition = False
+    for fv in encoded_fill_values:
+        condition |= data == fv
+    return np.where(condition, decoded_fill_value, data)
+
+
+class CFMaskCoder(VariableCoder):
+    """Mask or unmask fill values according to CF conventions."""
+
+    def encode(self, variable, name=None):
+        dims, data, attrs, encoding = unpack_for_encoding(variable)
+
+        if encoding.get('_FillValue') is not None:
+            fill_value = pop_to(encoding, attrs, '_FillValue', name=name)
+            if not pd.isnull(fill_value):
+                data = duck_array_ops.fillna(data, fill_value)
+            variable = Variable(dims, data, attrs, encoding)
+
+        if ('_FillValue' not in attrs and '_FillValue' not in encoding and
+                np.issubdtype(data.dtype, np.floating)):
+            attrs['_FillValue'] = data.dtype.type(np.nan)
+
+        return Variable(dims, data, attrs, encoding)
+
+    def decode(self, variable, name=None):
+        dims, data, attrs, encoding = unpack_for_decoding(variable)
+
+        if 'missing_value' in attrs:
+            # missing_value is deprecated, but we still want to support it as
+            # an alias for _FillValue.
+            if ('_FillValue' in attrs and
+                not utils.equivalent(attrs['_FillValue'],
+                                     attrs['missing_value'])):
+                raise ValueError("Conflicting _FillValue and missing_value "
+                                 "attrs on a variable {!r}: {} vs. {}\n\n"
+                                 "Consider opening the offending dataset "
+                                 "using decode_cf=False, correcting the "
+                                 "attrs and decoding explicitly using "
+                                 "xarray.decode_cf()."
+                                 .format(name, attrs['_FillValue'],
+                                         attrs['missing_value']))
+            attrs['_FillValue'] = attrs.pop('missing_value')
+
+        if '_FillValue' in attrs:
+            raw_fill_value = pop_to(attrs, encoding, '_FillValue', name=name)
+            encoded_fill_values = [
+                fv for fv in np.ravel(raw_fill_value) if not pd.isnull(fv)]
```

Comment:
Is it even possible to specify a sequence of fill values in a netcdf file?

Reply:
Apparently? This came from a pull request, though I refactored the logic here slightly.